dpnp/linalg/dpnp_utils_linalg.py

```diff
@@ -383,6 +385,78 @@ def _lu_factor(a, res_type):
         return (a_h, ipiv_h, dev_info_array)
 
 
+def _multi_dot(arrays, order, i, j, out=None):
+    """Actually do the multiplication with the given order."""
+    if i == j:
+        # the initial call with non-None out should never get here
+        assert out is None
+        return arrays[i]
+
+    return dpnp.dot(
+        _multi_dot(arrays, order, i, order[i, j]),
+        _multi_dot(arrays, order, order[i, j] + 1, j),
+        out=out,
+    )
+
+
+def _multi_dot_matrix_chain_order(n, arrays, return_costs=False):
+    """
+    Return a dpnp.ndarray that encodes the optimal order of mutiplications.
+
+    The optimal order array is then used by `_multi_dot()` to do the
+    multiplication.
+
+    Also return the cost matrix if `return_costs` is ``True``.
+
+    The implementation CLOSELY follows Cormen, "Introduction to Algorithms",
+    Chapter 15.2, p. 370-378.  Note that Cormen uses 1-based indices.
+
+        cost[i, j] = min([
+            cost[prefix] + cost[suffix] + cost_mult(prefix, suffix)
+            for k in range(i, j)])
+
+    """
+
+    # p stores the dimensions of the matrices
+    # Example for p: A_{10x100}, B_{100x5}, C_{5x50} --> p = [10, 100, 5, 50]
+    p = [1 if arrays[0].ndim == 0 else a.shape[0] for a in arrays]
+    p.append(1 if arrays[-1].ndim == 0 else arrays[-1].shape[1])
+    # m is a matrix of costs of the subproblems
+    # m[i,j]: min number of scalar multiplications needed to compute A_{i..j}
+    m = dpnp.zeros((n, n))
```

Comment:
need to allocate memory on the same queue and to cover the use case in `test_sycl_queue.py` and in `test_usm_type.py`

Reply:
Such a case is already covered in `test_sycl_queue.py` and `test_usm_type.py` (multiplication of five arrays).
Here, `m` and `s` are used to calculate `order` and all of them can be on a device different than input arrays.
For instance, the code can reads `order` on `gpu` and do the array dot on `cpu` according to `order`. 
Does it matter for performance if they are not on the same device?

Reply:
I wonder if it will lead to excess `asarray` call to migrate index on the same SYCL queue wher einput arrays reside, but I'm not 100% sure that it is a use case here.
Btw I believe it will be reasonable to check that all input arrays are on the same queue before starting dot computations.